app/jobs/verify_link_job.rb

```diff
@@ -0,0 +1,102 @@
+class VerifyLinkJob < ApplicationJob
+  queue_as :default
+
+  retry_on ActiveRecord::RecordNotFound, ActiveRecord::RecordInvalid, wait: :exponentially_longer, attempts: 3
+
+  ERRORS = (HTTP_ERRORS + [Faraday::Error, SocketError, SystemCallError, OpenSSL::SSL::SSLError]).freeze
+
+  class NotHTTPSError < StandardError; end
+  class LinkNotPresentError < StandardError; end
+  class HTTPResponseError < StandardError; end
+
+  discard_on NotHTTPSError do |job, _error|
+    job.record_failure
+  end
+
+  rescue_from LinkNotPresentError, HTTPResponseError, *ERRORS do |error|
+    logger.info "Linkback verification failed with error: #{error.message}", error: error, uri: link_verification.uri,
+      linkable: link_verification.linkable
+
+    link_verification.transaction do
+      record_failure
+      if should_retry?
+        retry_job(wait: 5.seconds * (3.5**link_verification.failures_since_last_verification.pred), error:)
+      else
+        instrument :retry_stopped, error: error
+      end
+    end
+  end
+
+  TIMEOUT_SEC = 5
+
+  def perform(link_verification:)
+    verify_link!(link_verification.uri, link_verification.linkable)
+    record_success
+  end
+
+  def link_verification
+    arguments.first.fetch(:link_verification)
+  end
+
+  def verify_link!(uri, linkable)
+    raise NotHTTPSError unless uri.start_with?("https://")
+
+    expected_href = linkable.linkable_verification_uri.to_s.downcase
+
+    response = get(uri)
+    raise HTTPResponseError, "Expected 200, got #{response.status}" unless response.status == 200
+    # TODO: body_with_limit, https://github.com/mastodon/mastodon/blob/33c8708a1ac7df363bf2bd74ab8fa2ed7168379c/app/lib/request.rb#L246
+    doc = Nokogiri::HTML5(response.body)
+
+    xpaths = [
+      # rel=me, what mastodon uses for profile link verification
+      '//a[contains(concat(" ", normalize-space(@rel), " "), " me ")]',
+      '//link[contains(concat(" ", normalize-space(@rel), " "), " me ")]',
+
+      # rel=rubygem
+      '//a[contains(concat(" ", normalize-space(@rel), " "), " rubygem ")]',
+      '//link[contains(concat(" ", normalize-space(@rel), " "), " rubygem ")]'
+    ]
+
+    if URI(uri).host == "github.com"
+      # github doesn't set a rel attribute on the URL added to a repo, so we have to use role=link instead
+      xpaths << '//a[contains(concat(" ", normalize-space(@role), " "), " link ")]'
+      xpaths << '//link[contains(concat(" ", normalize-space(@role), " "), " link ")]'
+    end
+
+    links = doc.xpath(xpaths.join("|"))
+
+    return if links.any? { |link| link["href"]&.downcase == expected_href }
+    raise LinkNotPresentError, "Expected #{expected_href} to be present in #{uri}"
```

Comment:
If you returned false and then abstracted the "failed" process into a method that is also called by rescue_from, you could avoid this exception for control flow. I would also have this normal non-success case not log anything because not finding a linkback to rubygems is not an error, it's just the normal state of things for lots of links.

Reply:
I found this easiest while trying to debug the behavior of the job